### DISCUSSION TITLE
CRUD web app: language, frameworks, learning curve for a beginner

### DISCUSSION THREAD

**Discussion by StevenKlapNL** [community/community#127403] [Programming Help]
### Body

Hi all!

I just want to pick your brains on programming language, frameworks, learning curve, etc. for someone who wants to give it a go in creating a web app, that someone is an absolute beginner.

To give you a bit of an intro, I’m a complete noob when it comes to web app development. But I do have a healthy interest in it, and like to think I have a basic understanding of programming. In the past I played around with Dreamweaver, Flash, Access, and by now watched hours of YouTube tutorials by some of the experts around here.

More recently I made some business tools in Excel, one of these tools I’d like to Web-ify so that it can easily be used and updated by my team.

Basic functions of the Excel:
- Register customers per country
- Register various contact persons per customer
- Register sales activities per customer, and done by who from my team (was it a call, or a meeting, bullet points of what was discussed, action points, etc. etc.)
- Register in which sales phase the customer is, based on sales activity results (8 phases: 1. Lead / 2. Connected / 3. Qualify / 4. Pitch / 5. Overcome Objections / 6. Close the sale / 7. Implementation / 8. Follow-up)
- Filter based on customer name, see the complete sales activity history in chronological order
- Sort customers based on region (seeing all customers in a particular city region)

What I’m looking for is to make a web app which is:
- Accessible for whomever is a registered user
- Sales colleagues can check the status of any particular customer
- Add customers / activities
- Add specific action points based on sales phase
- Add action points to an agenda
- Generate an overview of previous activities (handy when visiting a customer)
- Keep track of all data in a SQL database
- Admin panel, person working on country "A" should not be able to see data from country "B", while Admin should be able to see al
- Perhaps at a later stage generate some simple statistics
- So yes, basically a very CRUDe CRM

I know there are plenty open source examples around there, here, everywhere. However, that kills the joy of figuring things out myself. I’d like to build this myself, but have no clue where to start, which language, framework, database.

In the last couple of days I’ve played around with:
- Visual Studio 2022
- VS Code 2019
- MS SQL Server
- MySQL
- ASP.NET CORE MVC
- Vue.js
- Flutter
- Currently diving headfirst into Flask (as I understand, this is supposed to be an easy one, and a good steppingstone towards Django)

Looking forward hearing your tips, tricks and pointers!

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by javierpaniza** [comment]
Have a look at OpenXava:
https://github.com/openxava/openxava

**Reply by ebndev** [reply]
Hey @javierpaniza - thanks for the suggestion! Can you talk a bit more about why this would be useful?

**Comment by ebndev** [comment]
Hey @StevenKlapNL, thanks for being a part of the GitHub Community!

Since you mentioned not having much web-development experience, this seems like a pretty big undertaking so all of the pieces may not coming together immediately - but that shouldn't discourage you!

Since you know you'd like to use a SQL database, I would suggest learning SQL first, setting up your database, inputting test data, then making sure you can query all of the information that you would need.

From there, here's a tutorial on [How to Connect React JS With MySQL Database using Node.JS/Express.js](https://www.youtube.com/watch?v=Q3ixb1w-QaY). You'll need to have a strong understanding of JavaScript in order to be able to write code in React [front-end] and Node [back-end].

Node is great because there are a ton of packages out there that will do heavy lifting for you (i.e Passport handles user authentication).

If you're starting from the very beginning, here's a great series of videos on [Building CRUD Applications](https://youtu.be/jZ-kmmgi_d0?t=1861). They are using a NoSQL database though (MongoDB).
